src/dft/backends/mklcpu/commit.cpp

```diff
@@ -31,113 +31,161 @@
-            : detail::commit_impl<prec, dom>(queue, backend::mklcpu) {
-        DFT_ERROR status = DFT_NOTSET;
-        if (config_values.dimensions.size() == 1) {
-            status = DftiCreateDescriptor(&handle, get_precision(prec), get_domain(dom), 1,
-                                          config_values.dimensions[0]);
-        }
-        else {
-            status = DftiCreateDescriptor(&handle, get_precision(prec), get_domain(dom),
-                                          config_values.dimensions.size(),
-                                          config_values.dimensions.data());
-        }
-        if (status != DFTI_NO_ERROR) {
-            throw oneapi::mkl::exception(
-                "dft/backends/mklcpu", "commit",
-                "DftiCreateDescriptor failed with status: " + std::to_string(status));
-        }
+namespace detail {
+
+template <dft::detail::precision prec, dft::detail::domain dom>
+commit_derived_impl<prec, dom>::commit_derived_impl(
+    sycl::queue queue, const dft::detail::dft_values<prec, dom>& config_values)
+        : oneapi::mkl::dft::detail::commit_impl<prec, dom>(queue, backend::mklcpu) {
+    // create the descriptor once for the lifetime of the descriptor class
+    DFT_ERROR status[2] = { DFTI_BAD_DESCRIPTOR };
+
+    for (auto dir : { DIR::fwd, DIR::bwd }) {
+        if (config_values.dimensions.size() == 1)
+            status[dir] = DftiCreateDescriptor(&bidirection_handle[dir], mklcpu_prec, mklcpu_dom, 1,
+                                               config_values.dimensions[0]);
+        else
+            status[dir] = DftiCreateDescriptor(&bidirection_handle[dir], mklcpu_prec, mklcpu_dom,
+                                               config_values.dimensions.size(),
+                                               config_values.dimensions.data());
     }
 
-    void commit(const detail::dft_values<prec, dom>& config_values) override {
-        set_value(handle, config_values);
-        auto status = DftiCommitDescriptor(handle);
-        if (status != DFTI_NO_ERROR) {
-            throw oneapi::mkl::exception(
-                "dft/backends/mklcpu", "commit",
-                "DftiCommitDescriptor failed with status: " + std::to_string(status));
-        }
+    if (status[0] != DFTI_NO_ERROR && status[1] != DFTI_NO_ERROR) {
+        std::string err = std::string("DftiCreateDescriptor failed with status : ") +
+                          DftiErrorMessage(status[0]) + std::string(", ") +
+                          DftiErrorMessage(status[1]);
+        throw oneapi::mkl::exception("dft/backends/mklcpu", "create_descriptor", err);
     }
+}
 
-    virtual void* get_handle() noexcept override {
-        return handle;
-    }
+template <dft::detail::precision prec, dft::detail::domain dom>
+commit_derived_impl<prec, dom>::~commit_derived_impl() {
+    for (auto dir : { DIR::fwd, DIR::bwd })
+        DftiFreeDescriptor(&bidirection_handle[dir]);
```

Comment:
Check return error

Reply:
done in 86c1ec999fca920b2212e90b4dad63e56f367b10

Reply:
looks like this was reverted thereafter in [89edd4dcb64614b420bb835325b493b9bab10ba0](https://github.com/oneapi-src/oneMKL/pull/288/commits/89edd4dcb64614b420bb835325b493b9bab10ba0)

Reply:
clarified offline as well